plumbing/transport/http/common.go

```diff
@@ -260,23 +262,37 @@ func NewErr(r *http.Response) error {
 
 	switch r.StatusCode {
 	case http.StatusUnauthorized:
-		return transport.ErrAuthenticationRequired
+		return &Err{Response: r, WrappedError: transport.ErrAuthenticationRequired}
 	case http.StatusForbidden:
-		return transport.ErrAuthorizationFailed
+		return &Err{Response: r, WrappedError: transport.ErrAuthorizationFailed}
 	case http.StatusNotFound:
-		return transport.ErrRepositoryNotFound
+		return &Err{Response: r, WrappedError: transport.ErrRepositoryNotFound}
 	}
 
-	return plumbing.NewUnexpectedError(&Err{r})
+	return plumbing.NewUnexpectedError(&Err{Response: r})
 }
 
 // StatusCode returns the status code of the response
 func (e *Err) StatusCode() int {
 	return e.Response.StatusCode
 }
 
+// Body returns the body of the response
+func (e *Err) Body() io.ReadCloser {
+	return e.Response.Body
```

Comment:
At present, all `NewErr` calls ensure that  the body is closed, which is required in order to avoid resource leaks, as per docs:

> If the returned error is nil, the Response will contain a non-nil Body which the user is expected to close. If the Body is not both read to EOF and closed, the Client's underlying RoundTripper (typically Transport) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.

I am not sure there is a benefit of exposing a `Body()` that is already (or soon to be) closed.

Reply:
The function can read the body and store in local cache or add to the error message. 
I have another workaround by intercepting the response in custom round tripper, but I really don't like that solution

Reply:
> I am not sure there is a benefit of exposing a `Body()`

as I said we need it to analyze and print the full response, including the body. For example the GitLab can return this response: `403 Forbidden  - Your account has been blocked.` In the case of such response, we don't have information about the reason of 403 response, it can be, like in that example, the account is blocked, or invalid token/credentials.